File: 01_introduction_to_algorithms/Python/01_binary_search.py
def binary_search(list, item): # В переменных low high хранятся границы
    low = 0                    # той части списка, в которой выполняется поиск
    high = len(list) - 1

    while low <= high:          # Пока эта часть не сократиться до одного элемента...
        mid = (low + high) // 2 # ...проверяем средний элемент
        guess = list[mid]
        if guess == item:       # Значение найдено
            return mid
        elif guess > item:      # много
            high = mid -1
        else:                   # Мало
            low = mid + 1
    return None                 # Значение не существует

my_list = [1, 3, 5, 7, 9]       # массив для тестирования функции

print(binary_search(my_list, 3)) # 1 (нумерация элементов начинается с 0)
print(binary_search(my_list, -1)) # None (такого элемента нет в нашем списке)

File: 01_introduction_to_algorithms/JS/01_binary_search.js
"use strict";

function binary_search(list, item) {
    var low = 0;
    var high = list.length - 1;
    
    while (low <= high) {
        var mid = Math.floor((low + high) / 2);
        var guess = list[mid];

        if (guess === item) {
            return mid;
        } else if (guess > item) {
            high = mid - 1;
        } else {
            low = mid + 1;
        }
    }
    return null;
};

var my_list = [1, 3, 5, 7, 9];
console.log(binary_search(my_list, 3));
console.log(binary_search(my_list, -1));

alert('Я - JavaScript!');
